### PR TITLE
add @GeneratedValue annotation that can be used to autogenerate Id value on insert

### PR DESCRIPTION
**Pull request by Blackbaud-EricSlater**
this is a port of this PR from the old repo...
https://github.com/microsoft/spring-data-cosmosdb/pull/481

One thing to point out about this, given the discussion from the old PR, is that this is an entirely optional feature and changes no existing behavior. It simply adds an @GeneratedValue annotation that users may use to auto generate values for string type id fields and is consistent with the functionality in spring-data-mongo

### CODE REVIEW

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater thanks for the contribution, I will review it soon.

**Comment by kushagraThapar** [comment]
@kirankumarkolli @moderakh  - are we okay with auto generated ids on spring-data-cosmos side ?
This supports auto generating ids, which is very core spring data feature. The way it is implemented in this PR is: 
1. It is optional. 
2. It will only generate automatic Ids for the entities that this annotation gets applied on. 
3. It will not touch any other code base. 

However, since we stopped supporting this on Cosmos SDK, that's why we didn't include this feature in previous spring-data-cosmosdb SDK. 

Here is more discussion on this between @kirankumarkolli and the author earlier: https://github.com/microsoft/spring-data-cosmosdb/pull/481

**Comment by kirankumarkolli** [comment]
Personally I am not great fan of auto-generation of critical important data points like (PK & id).
But if similar such conventions/usages are popular in some tools/fx, opt-in why seems reasonable.

**Comment by moderakh** [comment]
I remember we had an issue with UUID generation in activityId causing log pauses due to lack of real randomness entropy. 
Do other spring plugin use auto generated UUID? I wonder if we should use fast uuid generation for this (similar to activityId in the SDK)

File: sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java [hunk 1]
```diff
@@ -189,6 +193,14 @@ public void setApplicationContext(ApplicationContext applicationContext) throws
         return toDomainObject(domainType, response.getItem());
     }
 
+    @SuppressWarnings("unchecked")
+    private <T> void generateIdIfNullAndAutoGenerationEnabled(T originalItem, Class<?> type) {
+        CosmosEntityInformation<?, ?> entityInfo = CosmosEntityInformation.getInstance(type);
+        if (entityInfo.shouldGenerateId() && ReflectionUtils.getField(entityInfo.getIdField(), originalItem) == null) {
+            ReflectionUtils.setField(entityInfo.getIdField(), originalItem, UUID.randomUUID().toString());
```
Comment (moderakh) [inline sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:200]: I remember we had an issue with UUID generation in activityId causing log pauses due to lack of real randomness entropy. 
Do other spring plugin use auto generated UUID? I wonder if we should use fast uuid generation for this (similar to activityId in the SDK)

**Comment by kushagraThapar** [comment]
> Personally I am not great fan of auto-generation of critical important data points like (PK & id).
> But if similar such conventions/usages are popular in some tools/fx, opt-in why seems reasonable.

Yes, opt-in auto-generation id is popular ask in spring-data world, that's why this makes sense just for id.

File: sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:200]: @moderakh - can you please point to the code where we auto generate activity id in the SDK code, may be we can implement something similar in spring if it helps with the performance.

Comment (moderakh) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:200]: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java#L438-L440

Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:200]: Thanks for the code @moderakh , however the above TimeBasedUUIDGenerator is being brought in by uuid external dependency which is currently in the banned dependency list. 
I also checked our V2 async code, and we use the randomUUID generator to generate UUIDs. 

In my opinion, it is not worth going through internal shading process for this. What are your thoughts?

Comment (moderakh) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:200]: One suggestion, write a for loop with UUID.randomUUID() and measure the time for each UUID generation, if the time looks good then go with UUID.randomUUID()

```java
int start = System.currentMillis();
for (int i = 0; i < 1_000_000; i++) {
  UUID.randomUUID();
}
int end = System.currentMillis();
System.out.println(end - start);
```

preferably run this on a linux VM where you don't have mouse movements, etc, as those generate random entropy which is used by /dev/random.

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/GenIdEntity.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.azure.spring.data.cosmos.core.mapping.GeneratedValue;
+import org.springframework.data.annotation.Id;
+
+import java.util.Objects;
+
+public class GenIdEntity {
+
+    @Id
+    @GeneratedValue
```
Comment (moderakh) [inline sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/GenIdEntity.java:13]: what happens if a field of non String type is annotated with `@ GeneratedValue` do we support that? or do we error out?

**Comment by Blackbaud-EricSlater** [comment]
added feature description and sample code to README

File: sdk/cosmos/azure-spring-data-cosmos-core/src/samples/java/com/azure/cosmos/GeneratedIdEntity.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.azure.cosmos;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos-core/src/samples/java/com/azure/cosmos/GeneratedIdEntity.java:1]: Please add Microsoft headers to this file.

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/GenIdEntity.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/GenIdEntity.java:13]: We error it out that `@GeneratedValue` can only be used on a string id.

File: sdk/cosmos/azure-spring-data-cosmos-core/src/samples/java/com/azure/cosmos/GeneratedIdEntity.java (hunk 3 above)
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos-core/src/samples/java/com/azure/cosmos/GeneratedIdEntity.java:3]: fixed

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - I think when you added the copyright headers in the GeneratedIdEntity sample file, you need to update the README.md file with updated line numbers - which I think should be 8 - 14 :)

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - I am seeing some more errors in spotbugs and checkstyles - can you please fix them ?

[ERROR] D:\a\1\s\sdk\cosmos\azure-spring-data-cosmos-core\src\samples\java\com\azure\cosmos\GeneratedIdEntity.java:12:19: Variable 'id' must be private and have accessor methods. [VisibilityModifier]

Also, once you fix this issue, you will also have to update the readme again with correct line numbers and the code snippet.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - to make sure everything passes on spotbugs, checkstyles and readme file, please run the below command locally to verify. 

```
mvn -e -Dgpg.skip -DskipTests -Dverify-readme -Dmaven.javadoc.skip=false -Dspotbugs.skip=false -Dcheckstyle.skip=false -Drevapi.skip=true -pl ,com.azure:azure-spring-data-cosmos-core,com.azure:azure-spring-data-cosmos-test -am clean verify
```

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - I am pretty sure the builds will fail, since you forgot to update the readme file with the latest change you made in the `GeneratedIdEntity.java` class. 

Please make sure readme.md is updated with whatever the class's contents are.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - I merged it, will fix it later, thanks for the contribution, really appreciate it.

**Pull request merged**